R/scalevariableset.R

```diff
@@ -0,0 +1,95 @@
+#' Scale a Displayr Variable Set
+#'
+#' Given a Displayr Variable Set or Q Question, performs a specified
+#' scaling on the data; first converting any categorical variables to
+#' numeric as necessary.
+#' @param data A Displayr Variable Set or Q Question: A numeric or
+#'     factor vector, or a data.frame containing either all numeric
+#'     columns or all factor columns.
+#' @param type Character string specifying the type of scaling to
+#'     perform; one of \code{"standardize"} (the default),
+#'     \code{"center"}, or \code{"unit"}.
+#' @param within.case Logical; should the scaling be performed with
+#'     case (rows) or variables (columns). An error is through if
+#'     \code{TRUE} and \code{data} contains only one variable.
+#' @return A numeric vector if \code{data} contains only a single
+#'     variable or a data.frame containing the scaled values.
+#' @seealso \code{\link{scale}}
+#' @export
+ScaleVariableSet <- function(
+                             data,
+                             type = c("standardize", "center", "unit"),
+                             within.case = FALSE)
+{
+    type <- match.arg(type)
+
+    if (within.case && NCOL(data) == 1L)
+        stop(shQuote("data"), "contains only one variable, scaling within case is ",
+             "not meaningful.")
+
+    is.categorical <- is.factor(data) || (is.data.frame(data) && is.factor(data[[1]]))
+    if (is.categorical)
+        data.with.values <- replaceFactorsWithValues(data)
+    else
+        data.with.values <- as.matrix(data)
+        ## may need to drop NET
+        ## if (NCOL(data.with.values) > 1)
+        ## data.with.values <- data.with.values[, !grepl("^NET$|^SUM$", colnames(data.with.values))]
+
+
+    if (within.case)
+        data.with.values <- t(data.with.values)
+
+    if (type == "unit")
+        out <- apply(data.with.values, 2,
+                       function(x)
+                       {
+                           min.x <- min(x, na.rm = TRUE)
+                           max.x <- max(x, na.rm = TRUE)
+                           return((x - min.x)/(max.x - min.x))
+                       })
+    else
+        out <- scale(data.with.values, scale = type != "center")
+
+    if (within.case)
+        out <- t(out)
+    else
+        out <- drop(out)
```

Comment:
Is it safe to use drop here? Won't that mess with the dimension of the data and cause it to be the wrong size to fit in the datasets tree?

Reply:
No, it does not seem to cause any issues in my tests.